README.md
<div align="center">
  <br><br>
  <img src="media/configpilot-logo.png" height="80" width="185" alt="ConfigPilot">
  <br><br>
  <p>ConfigPilot is a lightweight and powerful configuration parser for Python that automates checks and typecasting.<br>Do everything you did in fewer lines of code!</p>
  <br>
</div>

## Features
- Simple and concise syntax.
- Lightweight and fast.
- Automatic type casting.
- Automation of value checks.
- Support for primitive Python types and user-defined types.
- Support for multi-line values and lists.
- No dependency.

## Installation

The recommended way to install ConfigPilot is to use `pip3`:

```shell
$ pip3 install configpilot
```

ConfigPilot requires Python 3.6 or later.

Import ConfigPilot into your project:

```python
from configpilot import ConfigPilot, OptionSpec

# Exceptions (optional)
from configpilot import ConfigPilotError, NoSectionError, \
                        NoOptionError, IllegalValueError
```

## Examples
### 1. A simple example
#### Configuration file
```ini
# This is a comment
[author]
 name      = 'John DOE' # This is an inline comment
 age       = 27
 github    = 'https://github.com'
 skills    = 'sleep well on airplanes'
             'have a terrific family recipe for brownies'
             'always up for dessert'
```
*Quotes are optional for strings (unless you put special characters).*

For this first example, we want to retrieve:
- The `name` as a string.
- The `age` as an integer between `0` and `100`.
- The `github` option as a string.
- The `skills` as a list of strings.

To achieve this, we have to:
- Define the desired file structure.
- Instantiate a `ConfigPilot` object and indicate the structure of the file.
- Read the file and check that it does not contain any errors in terms of format or content.
- Retrieve values.

#### Python code
```python
options = [
    OptionSpec(
        section='author',
        option='name'
    ),

    OptionSpec(
        section='author',
        option='age',
        allowed=range(0, 100),
        type=int
    ),

    OptionSpec(
        section='author',
        option='github'
    ),

    OptionSpec(
        section='author',
        option='skills',
        type=[str]
    )
]

config = ConfigPilot()
config.register(*options)
config.read('/path/file.conf')

if not config.is_opened:
    print('Error: unable to read the configuration file.')
    exit(1)

if config.errors:
    print('Error: some options are incorrect.')
    exit(1)

name = config.author.name      # 'John DOE'
age = config.author.age        # 27
github = config.author.github  # 'https://github.com'
skills = config.author.skills  # ['sleep well on airplanes',
                               #  'have a terrific family recipe for brownies',
                               #  'always up for dessert']

# Alternative syntax
name = config['author']['name']
age = config['author']['age']
github = config['author']['github']
skills = config['author']['skills']
```

### 2. Use more complex types
#### Configuration file
```ini
[general]
 mode:       master
 interface:  ens33
 port:       5000
 initDelay:  0.5

[logging]
 enabled:    false

[nodes]
 slaves:     10.0.0.1
             10.0.0.2
             10.0.0.3
```

What we want to retrieve:
- The `mode` option as a string. Two values will be possible: `master` or `slave`.
- The `interface` as a string. If the option is not specified, we will use the default value `ens33`.
- The `port` as an integer between `1024` and `49151`. The default value will be `4000`.
- The `initDelay` option as a float. The default value will be `0.0`.
- The `enabled` option, from the `logging` section, as a boolean.
- The `slaves`, from the `nodes` section, as a list of `IPv4Address` (from the `ipaddress` module).

#### Python code
```python
from ipaddress import IPv4Address

options = [
    OptionSpec(
        section='general',
        option='mode',
        allowed=('master', 'slave')
    ),

    OptionSpec(
        section='general',
        option='interface',
        default='ens33'
    ),

    OptionSpec(
        section='general',
        option='port',
        allowed=range(1024, 49151),
        default=4000,
        type=int
    ),

    OptionSpec(
        section='general',
        option='initDelay',
        default=0.0,
        type=float
    ),

    OptionSpec(
        section='logging',
        option='enabled',
        type=bool
    ),

    OptionSpec(
        section='nodes',
        option='slaves',
        type=[IPv4Address]
    )
]

config = ConfigPilot()
config.register(*options)
config.read('/path/file.conf')

if not config.is_opened:
    print('Error: unable to read the configuration file.')
    exit(1)

if config.errors:
    print('Error: some options are incorrect.')
    exit(1)

mode = config.general.mode             # 'master'
interface = config.general.interface   # 'ens33'
port = config.general.port             # 5000
init_delay = config.general.initDelay  # 0.5
logs_enabled = config.logging.enabled  # False
slaves = config.nodes.slaves           # [IPv4Address('10.0.0.1'),
                                       #  IPv4Address('10.0.0.2'),
                                       #  IPv4Address('10.0.0.3')]
```

### 3. Use functions and lambda functions
#### Configuration file
```ini
[boot]
 hexCode:    0x2A

[statistics]
 lastBoot:   2020-02-01 10:27:00
 lastCrash:  2019-12-10 09:00:00
```

What we want to retrieve:
- The `hexCode` option as an integer (base 16).
- The `lastBoot` option as a `datetime` object.
- The `lastCrash` option as a `datetime` object.

We cannot set the `type` parameter of the `OptionSpec` objects to `datetime` because the constructor of `datetime` expects several parameters. The values contained in the configuration file are strings with a specific format. So, we have to process these data with a dedicated function.

#### Python code
```python
from datetime import datetime


def string_to_datetime(value):
    '''
    Cast a string to a datetime object.

    '''
    # Do not handle any exceptions that can be raised in this function.
    # They are processed by ConfigPilot: the option, which called the
    # function, is considered wrong if an exception is thrown.
    return datetime.strptime(value, '%Y-%m-%d %H:%M:%S')


options = [
    OptionSpec(
        section='boot',
        option='hexCode',
        type=lambda x: int(x, 16)
    ),

    OptionSpec(
        section='statistics',
        option='lastBoot',
        type=string_to_datetime
    ),

    OptionSpec(
        section='statistics',
        option='lastCrash',
        type=string_to_datetime
    )
]

config = ConfigPilot()
config.register(*options)
config.read('/path/file.conf')

if not config.is_opened:
    print('Error: unable to read the configuration file.')
    exit(1)

if config.errors:
    print('Error: some options are incorrect.')
    exit(1)

boot_hex_code = config.boot.hexCode       # 42
last_boot = config.statistics.lastBoot    # datetime.datetime(2020, 2, 1, 10, 27)
last_crash = config.statistics.lastCrash  # datetime.datetime(2019, 12, 10, 9, 0)
```

## Classes
### OptionSpec
A user-created object that represents the constraints that an option must meet to be considered valid.

#### Definition
```python
OptionSpec(section, option, allowed=None, default=None, type=str)
```

#### Parameters / Getters
- `section`

  The name of a section in the file.

  - Type: `str`

- `option`

  The name of an option in the specified section.

  - Type: `str`

- `allowed`

  The list or range of allowed values.

  - Type: `object that supports the 'in' operator (membership)`
  - Default: `None`

- `default`

  The default value of the option if it does not exist.<br>
  Must be an object of the same type as the value obtained after the cast (see the `type` parameter).

  - Type: `object`
  - Default: `None`

- `type`

  The expected value type for this option.<br>
  Set it to `int`, `float`, `bool`, `str` (default) or any other type of object.<br>
  If you expect a list of values, use instead `[int]`, `[float]`, `[bool]`, `[str]` (equivalent of `list`) or even `[MyClass]`.

  - Type: `type` or `list`
  - Default: `str`

### ConfigPilot
#### Definition
```python
ConfigPilot()
```

#### Methods
- `register(*specifications)`

  Register one or several specifications. You can call this method multiple times.<br>
  Each option in the configuration file must have its own specification. Call the `read` method next.

  - `*specifications` parameter: one or several `OptionSpec`.

- `read(filename, encoding='utf-8')`

  Read and parse a configuration file according to the registered specifications.

  - `filename` parameter: the name of the configuration file to read.
  - `encoding` parameter: the name of the encoding used to decode the file. The default encoding is `UTF-8`.

#### Getters
- `filename`

  The name of the last opened file.

  - Type: `str`

- `is_opened`

  Return a boolean that indicates whether the file is opened or not.

  - Type: `bool`

- `errors`

  Return a dictionary containing sections and options that do not meet the specifications.

  - Type: `dict`

## Contributing

Comments and enhancements are welcome.

All development is done on [GitHub](https://github.com/ValentinBELYN/configpilot). Use [Issues](https://github.com/ValentinBELYN/configpilot/issues) to report problems and submit feature requests. Please include a minimal example that reproduces the bug.

## Donate

ConfigPilot is completely free and open source. It has been fully developed on my free time. If you enjoy it, please consider donating to support the development.

- [:tada: Donate via PayPal](https://paypal.me/ValentinBELYN)

## License

Copyright 2017-2020 Valentin BELYN.

Code released under the MIT license. See the [LICENSE](LICENSE) for details.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [v1.1.0](https://github.com/ValentinBELYN/configpilot/releases/tag/v1.1.0) - 2020-05-19
- Add the ability to modify the default encoding.
- Fix a bug related to the default encoding of operating systems.
- Delete old development features.

## [v1.0.2](https://github.com/ValentinBELYN/configpilot/releases/tag/v1.0.2) - 2020-05-10
- Fix a bug affecting default values.

## [v1.0.1](https://github.com/ValentinBELYN/configpilot/releases/tag/v1.0.1) - 2020-01-02
- Add a logo to the project.
- Update the ConfigPilot package on PyPI.
- Documentation improvements.

## [v1.0.0](https://github.com/ValentinBELYN/configpilot/releases/tag/v1.0.0) - 2019-12-26
- :tada: Initial release.


configpilot/utils.py
'''
    configpilot
    ~~~~~~~~~~~

        https://github.com/ValentinBELYN/configpilot

    :copyright: Copyright 2017-2020 Valentin BELYN.
    :license: MIT, see the LICENSE for details.

    ~~~~~~~~~~~

    Permission is hereby granted, free of charge, to any person
    obtaining a copy of this software and associated documentation
    files (the 'Software'), to deal in the Software without
    restriction, including without limitation the rights to use, copy,
    modify, merge, publish, distribute, sublicense, and/or sell copies
    of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be
    included in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
    EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
    OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
    NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
    HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
    WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE
    OR OTHER DEALINGS IN THE SOFTWARE.
'''


def cast_to_boolean(value):
    '''
    Cast a string to a boolean. Raise a TypeError exception if the
    given value cannot be casted.

    Case insensitive.

    '''
    value = value.lower()

    boolean_states = {
        'true': True,
        'false': False
    }

    if value in boolean_states:
        return boolean_states[value]

    raise TypeError


def cast_value(value, expected_type):
    '''
    Cast a value (preferably a string) in the specified type. Raise an
    exception (depending on the type passed in parameters) if the value
    cannot be casted.

    '''
    if expected_type is bool:
        expected_type = cast_to_boolean

    return expected_type(value)


def cast_values(values, expected_type):
    '''
    Cast the values (preferably strings) of a list in the specified
    type. Raise an exception (depending on the type passed in
    parameters) if a value cannot be casted.

    '''
    if expected_type is bool:
        expected_type = cast_to_boolean

    return [
        expected_type(value)
        for value in values
    ]


def magic_cast(*values, expected_type):
    '''
    Cast any value (preferably strings) in the specified type, whether
    it is a single value or a list of values. Raise an exception
    (depending on the type passed in parameters) if a value cannot be
    casted.

    Usage::

        # Cast strings to a list of integers
        >>> magic_cast('1', '2', '3', expected_type=[int])
        [1, 2, 3]

        # Cast a string to an integer
        >>> magic_cast('1', expected_type=int)
        1

        # Cast a string to a boolean
        >>> magic_cast('true', expected_type=bool)
        True

    '''
    if (expected_type is list or
        expected_type == []):
        expected_type = [str]

    if isinstance(expected_type, list):
        return cast_values(values, expected_type[0])

    if len(values) == 1:
        return cast_value(values[0], expected_type)

    raise TypeError


configpilot/models.py
'''
    configpilot
    ~~~~~~~~~~~

        https://github.com/ValentinBELYN/configpilot

    :copyright: Copyright 2017-2020 Valentin BELYN.
    :license: MIT, see the LICENSE for details.

    ~~~~~~~~~~~

    Permission is hereby granted, free of charge, to any person
    obtaining a copy of this software and associated documentation
    files (the 'Software'), to deal in the Software without
    restriction, including without limitation the rights to use, copy,
    modify, merge, publish, distribute, sublicense, and/or sell copies
    of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be
    included in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
    EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
    OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
    NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
    HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
    WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE
    OR OTHER DEALINGS IN THE SOFTWARE.
'''

from .exceptions import NoOptionError, IllegalValueError


class Section:
    '''
    A class that represents a section.
    Generated by a ConfigPilot object.

    :type name: str
    :param name: The name of the section.

    '''
    def __init__(self, name):
        self._name = name
        self._options = {}

    def __repr__(self):
        return f'<Section [{self._name}]>'

    def __len__(self):
        '''
        Return the number of options contained in the section.

        '''
        return len(self._options)

    def __getattr__(self, name):
        '''
        Get the value of the specified option.

        :raises NoOptionError: If the option is not found in the
            configuration file.
        :raises IllegalValueError: If the option does not meet the
            constraints defined in its specification.

        '''
        if name[:2] == '__':
            return super().__getattr__(name)

        return self[name]

    def __getitem__(self, key):
        '''
        Get the value of the specified option.

        :raises NoOptionError: If the option is not found in the
            configuration file.
        :raises IllegalValueError: If the option does not meet the
            constraints defined in its specification.

        '''
        if key not in self._options:
            raise NoOptionError(self._name, key)

        return self._options[key].value

    def add(self, option):
        '''
        Add an option to the section.

        '''
        self._options[option.name] = option


class Option:
    '''
    A class that represents an option.
    Generated by a ConfigPilot object.

    :type name: str
    :param name: The name of the option.

    :type value: object
    :param value: The (casted) value of the option.

    :type specification: OptionSpec
    :param specification: The specification corresponding to the
        option.

    '''
    def __init__(self, name, value, specification):
        self._name = name
        self._value = value
        self._specification = specification

    def __repr__(self):
        return f'<Option [{self._name}]>'

    @property
    def name(self):
        '''
        The name of the option.

        '''
        return self._name

    @property
    def value(self):
        '''
        The (casted) value of the option.

        :raises IllegalValueError: If the option does not meet the
            constraints defined in its specification.

        '''
        if self._value is None:
            raise IllegalValueError(
                section=self._specification.section,
                option=self._name)

        return self._value

    @property
    def specification(self):
        '''
        The specification corresponding to the option.

        '''
        return self._specification


class OptionSpec:
    '''
    A user-created object that represents the constraints that an
    option must meet to be considered valid.

    :type section: str
    :param section: The name of a section in the file.

    :type option: str
    :param option: The name of an option in the specified section.

    :type allowed: object that supports the 'in' operator (membership).
    :param allowed: (Optional) The list or range of allowed values.

    :type default: object
    :param default: (Optional) The default value of the option if it
        does not exist. Must be an object of the same type as the value
        obtained after the cast (see the 'type' parameter).

    :type type: type or list
    :param type: (Optional) The expected value type for this option.
        Set it to 'int', 'float', 'bool', 'str' (default) or any other
        type of object. If you expect a list of values, use instead
        '[int]', '[float]', '[bool]', '[str]' (equivalent of 'list') or
        even '[MyClass]'.

    '''
    def __init__(self, section, option, allowed=None, default=None,
            type=str):

        self._section = section
        self._option = option
        self._allowed = allowed
        self._default = default
        self._type = type

    def __repr__(self):
        return f'<OptionSpec [{self._section}:{self._option}]>'

    @property
    def section(self):
        '''
        The name of a section in the file.

        '''
        return self._section

    @property
    def option(self):
        '''
        The name of an option in the specified section.

        '''
        return self._option

    @property
    def allowed(self):
        '''
        The list or range of allowed values.

        '''
        return self._allowed

    @property
    def default(self):
        '''
        The default value of the option if it does not exist.

        '''
        return self._default

    @property
    def type(self):
        '''
        The expected value type for this option.

        '''
        return self._type


configpilot/__init__.py
'''
    configpilot
    ~~~~~~~~~~~

        https://github.com/ValentinBELYN/configpilot

    :copyright: Copyright 2017-2020 Valentin BELYN.
    :license: MIT, see the LICENSE for details.

    ~~~~~~~~~~~

    Permission is hereby granted, free of charge, to any person
    obtaining a copy of this software and associated documentation
    files (the 'Software'), to deal in the Software without
    restriction, including without limitation the rights to use, copy,
    modify, merge, publish, distribute, sublicense, and/or sell copies
    of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be
    included in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
    EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
    OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
    NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
    HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
    WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE
    OR OTHER DEALINGS IN THE SOFTWARE.
'''

from .pilot import ConfigPilot
from .models import OptionSpec
from .exceptions import *


__author__    = 'Valentin BELYN'
__copyright__ = 'Copyright 2017-2020 Valentin BELYN'
__license__   = 'MIT License'

__version__   = '1.1'
__build__     = '200519'


__all__ = [
    'ConfigPilot',
    'ConfigPilotError',
    'NoSectionError',
    'NoOptionError',
    'IllegalValueError',
    'OptionSpec'
]


configpilot/exceptions.py
'''
    configpilot
    ~~~~~~~~~~~

        https://github.com/ValentinBELYN/configpilot

    :copyright: Copyright 2017-2020 Valentin BELYN.
    :license: MIT, see the LICENSE for details.

    ~~~~~~~~~~~

    Permission is hereby granted, free of charge, to any person
    obtaining a copy of this software and associated documentation
    files (the 'Software'), to deal in the Software without
    restriction, including without limitation the rights to use, copy,
    modify, merge, publish, distribute, sublicense, and/or sell copies
    of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be
    included in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
    EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
    OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
    NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
    HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
    WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE
    OR OTHER DEALINGS IN THE SOFTWARE.
'''


class ParseError(Exception):
    '''
    Base class for parsing exceptions.

    '''


class MissingSectionHeaderError(ParseError):
    '''
    Raised when a configuration file does not begin with a section.

    '''
    def __init__(self):
        message = 'The configuration file contains no section headers'
        super().__init__(message)


class MissingOptionError(ParseError):
    '''
    Raised when a section does not begin with an option.

    '''
    def __init__(self, section):
        message = f'The \'{section}\' section does not begin with ' \
                   'an option'
        super().__init__(message)


class MissingValueError(ParseError):
    '''
    Raised when no value is assigned to an option.

    '''
    def __init__(self, section, option):
        message = f'No value found for the \'{option}\' option in ' \
                  f'the \'{section}\' section'
        super().__init__(message)


class DuplicateSectionError(ParseError):
    '''
    Raised when the same section is defined multiple times in a
    configuration file.

    '''
    def __init__(self, section):
        message = f'The \'{section}\' section already exists'
        super().__init__(message)


class DuplicateOptionError(ParseError):
    '''
    Raised when the same option is defined multiple times in a section.

    '''
    def __init__(self, section, option):
        message = f'The \'{option}\' option already exists in ' \
                  f'the \'{section}\' section'
        super().__init__(message)


class ConfigPilotError(Exception):
    '''
    Base class for ConfigPilot exceptions.

    '''


class NoSectionError(ConfigPilotError):
    '''
    Raised when the requested section is not found in the configuration
    file.

    '''
    def __init__(self, section):
        message = f'The \'{section}\' section was not found'
        super().__init__(message)


class NoOptionError(ConfigPilotError):
    '''
    Raised when the requested option is not found in the configuration
    file.

    '''
    def __init__(self, section, option):
        message = f'The \'{option}\' option (from the \'{section}\' ' \
                   'section) was not found'
        super().__init__(message)


class IllegalValueError(ConfigPilotError):
    '''
    Raised when the requested option does not meet the constraints
    defined by the developer.

    '''
    def __init__(self, section, option):
        message = f'The \'{option}\' option (from the \'{section}\' ' \
                   'section) does not meet the constraints defined ' \
                   'by the OptionSpec object'
        super().__init__(message)


configpilot/pilot.py
'''
    configpilot
    ~~~~~~~~~~~

        https://github.com/ValentinBELYN/configpilot

    :copyright: Copyright 2017-2020 Valentin BELYN.
    :license: MIT, see the LICENSE for details.

    ~~~~~~~~~~~

    Permission is hereby granted, free of charge, to any person
    obtaining a copy of this software and associated documentation
    files (the 'Software'), to deal in the Software without
    restriction, including without limitation the rights to use, copy,
    modify, merge, publish, distribute, sublicense, and/or sell copies
    of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be
    included in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
    EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
    OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
    NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
    HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
    WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE
    OR OTHER DEALINGS IN THE SOFTWARE.
'''

from .parsers import INIParser
from .models import Section, Option
from .exceptions import ParseError, NoSectionError
from .utils import magic_cast


class ConfigPilot:
    '''
    A lightweight and powerful configuration parser for Python that
    automates checks and typecasting.

    Do everything you did in fewer lines of code.

    Usage::

        # # file.conf
        #
        # [general]
        #  mode      = master
        #  interface = ens33
        #  port      = 5000
        #
        # [nodes]
        #  slaves    = 10.0.0.1
        #              10.0.0.2
        #              10.0.0.3

        options = [
            OptionSpec(
                section='general',
                option='mode',
                allowed=('master', 'slave')
            ),

            OptionSpec(
                section='general',
                option='interface',
                default='ens33'
            ),

            OptionSpec(
                section='general',
                option='port',
                allowed=range(1024, 49151),
                default=4000,
                type=int
            ),

            OptionSpec(
                section='nodes',
                option='slaves',
                type=[IPv4Address]
            )
        ]

        config = ConfigPilot()
        config.register(*options)
        config.read('/path/file.conf')

        if not config.is_opened:
            print('Error: unable to read the configuration file.')
            exit(1)

        if config.errors:
            print('Error: some options are incorrect.')
            exit(1)

        mode = config.general.mode
        interface = config.general.interface
        port = config.general.port
        slaves = config.nodes.slaves

        # Alternative syntax
        mode = config['general']['mode']

    See the OptionSpec class for details.

    '''
    def __init__(self):
        self._filename = None
        self._is_opened = False
        self._specifications = []
        self._config = {}
        self._errors = {}

    def __repr__(self):
        return f'<ConfigPilot [{self._filename}]>'

    def __len__(self):
        '''
        Return the number of sections contained in the file.

        '''
        return len(self._config)

    def __getattr__(self, name):
        '''
        Get the specified section.

        :raises NoSectionError: If the section is not found in the
            configuration file.

        '''
        if name[:2] == '__':
            return super().__getattr__(name)

        return self[name]

    def __getitem__(self, key):
        '''
        Get the specified section.

        :raises NoSectionError: If the section is not found in the
            configuration file.

        '''
        if key not in self._config:
            raise NoSectionError(key)

        return self._config[key]

    def _get_section(self, name):
        if name not in self._config:
            self._config[name] = Section(name)

        return self._config[name]

    def _add_error(self, section, option):
        if section not in self._errors:
            self._errors[section] = []

        self._errors[section].append(option)

    def register(self, *specifications):
        '''
        Register one or several specifications. You can call this
        method multiple times.

        Each option in the configuration file must have its own
        specification. Call the 'read' method next.

        '''
        self._specifications.extend(specifications)

    def read(self, filename, encoding='utf-8'):
        '''
        Read and parse a configuration file according to the registered
        specifications.

        :type filename: str
        :param filename: The path of the file to read.

        :type encoding: str
        :param encoding: (Optional) The name of the encoding used to
            decode the file. The default encoding is UTF-8.

        :rtype: bool
        :returns: A boolean that indicates whether the file is opened
            or not.

        '''
        parser = INIParser()
        self._filename = filename
        self._is_opened = False
        self._config = {}
        self._errors = {}

        try:
            parser.read(filename, encoding)

        except (OSError, ParseError):
            return self._is_opened

        for spec in self._specifications:
            # We get the value(s) corresponding to the specification.
            # This method always returns the value(s) in a list.
            values = parser.get(
                section=spec.section,
                option=spec.option)

            # If the configuration file contains a value, we cast it
            # into the desired type and we verify that it is within the
            # specified range.
            if values:
                try:
                    casted_value = magic_cast(
                        *values,
                        expected_type=spec.type)

                    if (spec.allowed and
                        casted_value not in spec.allowed):
                        raise ValueError

                except Exception:
                    casted_value = None

                    self._add_error(
                        section=spec.section,
                        option=spec.option)

            # If no option is provided in the file and a default value
            # is set, we use this value.
            elif spec.default is not None:
                casted_value = spec.default

            # If no option is provided and no default value is defined,
            # we add an error to the option.
            else:
                self._add_error(
                    section=spec.section,
                    option=spec.option)
                continue

            option = Option(
                name=spec.option,
                value=casted_value,
                specification=spec)

            section = self._get_section(spec.section)
            section.add(option)

        self._is_opened = True
        return self._is_opened

    @property
    def filename(self):
        '''
        The name of the last opened file.

        '''
        return self._filename

    @property
    def is_opened(self):
        '''
        Return a boolean that indicates whether the file is opened or
        not.

        '''
        return self._is_opened

    @property
    def errors(self):
        '''
        Return a dictionary containing sections and options that do not
        meet the specifications.

        '''
        return self._errors
